src/coprocessor/dag/rpn_expr/impl_cast.rs

```diff
@@ -163,33 +163,32 @@ pub fn cast_int_as_decimal(
     }
 }
 
-macro_rules! cast_as_integer {
-    ($ty:ty, $as_int_fn:ident) => {
-        cast_as_integer!($ty, $as_int_fn, val);
-    };
-    ($ty:ty, $as_int_fn:ident, $expr:expr) => {
-        #[rpn_fn(capture = [ctx])]
-        #[inline]
-        pub fn $as_int_fn(ctx: &mut EvalContext, val: &Option<$ty>) -> Result<Option<i64>> {
-            match val {
-                None => Ok(None),
-                Some(val) => {
-                    let val = $expr.to_int(ctx, FieldTypeTp::LongLong)?;
-                    Ok(Some(val))
-                }
-            }
+#[rpn_fn(capture = [ctx])]
+#[inline]
+fn cast_any_to_int<T: ToInt + Evaluable>(
+    ctx: &mut EvalContext,
+    val: &Option<T>,
+) -> Result<Option<i64>> {
+    match val {
+        None => Ok(None),
+        Some(val) => {
+            let val = val.to_int(ctx, FieldTypeTp::LongLong)?;
```

Comment:
should `FieldTypeTp::LongLong` be `extra.ret_field_type.tp()`?

Reply:
The TiDB used `FieldTypeTp::LongLong`.

Reply:
OMG